### PR TITLE
Chronopost carrier integration

### PR DESCRIPTION
**Pull request by Ftayri**
karrio.chronopost

### CODE REVIEW

**Comment by danh91** [comment]
Hi @Ftayri,

I took a deeper look at what you have here and made some improvements to simplify the integration.
You are on the right track.

I will leave a couple of comments around and we can schedule a call to walk through edge cases.

File: schemas/chronopost/generate.sh
Comment (danh91) [inline schemas/chronopost/generate.sh:7]: I looked at the documentation and for the full integration of the shipping service, you will also need the QuickCost schema to parse the rating responses

https://ws.chronopost.fr/quickcost-cxf/QuickcostServiceWS?wsdl

Comment (danh91) [reply schemas/chronopost/generate.sh:7]: Actually, you might as well generate the schema types for TrackingService since the response for cancel shipment (cancelSkyBill) response schema will be required

https://ws.chronopost.fr/tracking-cxf/TrackingServiceWS?wsdl

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:29]: You will need to implement the API request here as per the QuickCost service documentation

```
https://ws.chronopost.fr/quickcost-cxf/QuickcostServiceWS/quickCost?
accountNumber=[accountNumber]&password=[password]&depCode=28500&arrCode=75001&weight=1&produ
ctCode=1&type=M
```

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:34]: I don't think the `soapaction` will be necessary for Chronopost in general

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:42]:
```suggestion
```

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:34]:
```suggestion
```

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/cancel.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/shipment/cancel.py:53]:
```suggestion
```

This is won't be necessary anymore

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py:110]: Is this required or we can keep it `None`

File: sdk/extensions/chronopost/karrio/providers/chronopost/units.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/units.py:39]: I simplified the definition of shipping options.
Please merge main to use the new approach

**Review by danh91 — CHANGES_REQUESTED** [review]
It's looking good overall, I will recommend you merge the latest `main` and leverage the new `cli` and helpers to move faster.

I am updating the documentation now to add the new `cli` and the improvements.

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py
Comment (Ftayri) [reply sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py:110]: Apparently it is required, the service would fail if I don't include it in the envelope.

**Comment by Ftayri** [comment]
Hi @danh91 ,
I updated the code to match the new structure. I also added the new schemas to implement cancelling shipments and fetching rates. For fetching rates I decided to use `calculateProdcuts` from the QuickCost webservice to fetch the rates. `quickCost` needs the service code in the request in order to fetch its rate and the rate of other services for some reason.
I'm also not sure if parsing responses and extracting details in both create.py and rate.py are well implemented.

**Comment by danh91** [comment]
No problem. I will take a look shortly and give you some guidance for the next steps.

File: .vscode/settings.json
Comment (danh91) [inline .vscode/settings.json:17]: Instead of this. Please add the package to the dev dependency

`requirements.sdk.dev.txt`
```txt
# Schemas update
-e ./schemas/chronopost
```
same for `requirements.server.dev.txt`
```txt
# Schemas update
-e ./schemas/chronopost
```

File: schemas/chronopost/schemas/QuickcostServiceWS.xml
Comment (danh91) [inline schemas/chronopost/schemas/QuickcostServiceWS.xml:1]: Please rename this files into `.xsd` instead of `.xml`

File: schemas/chronopost/schemas/ShippingServiceWS.xml
Comment (danh91) [inline schemas/chronopost/schemas/ShippingServiceWS.xml:1]: Please rename this files into `.xsd` instead of `.xml`

File: schemas/chronopost/schemas/TrackingServiceWS.xml
Comment (danh91) [inline schemas/chronopost/schemas/TrackingServiceWS.xml:1]: Please rename this files into `.xsd` instead of `.xml`

File: sdk/extensions/chronopost/karrio/mappers/chronopost/mapper.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/mapper.py:11]:
```suggestion
```

We will need to implement the chronopost-specific rate request and response so we can remove this import.

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py
Comment (danh91) [reply sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py:96]: No problem. What we can do is we can declare an option for that so it can be optionally customized but we will keep a default value as you did.

By any chance, is there another option that is more neutral instead of `M`?

File: sdk/extensions/chronopost/karrio/providers/chronopost/units.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/units.py:40]: That's good

**Review by danh91 — CHANGES_REQUESTED** [review]
Overall, You've made some good progress.
schedule a time here for us to do some pair programming and wrap it up
https://calendly.com/karrioapi/chat

**Comment by Ftayri** [comment]
@danh91 I managed to run the shipment integration tests, fixed a bunch of issues but I couldn't solve 2 problems. The first one is that the label isn't parsed for some reason. As for the second problem, cancelling shipment returns an error with code 4 saying document is empty.
Both problematic responses are `ParsedShipmentResponse` and `ParsedCancelShipmentResponse` in shipment.py if you want to look at them.

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py:29]:
```suggestion
    shipment_node = lib.find_element("resultMultiParcelValue", response, first=True)
    shipment = (
        _extract_details(result, settings)
        if shipment_node is not None else None
    )
```

I think that is the reason why the shipment parsing is failing.

**Review by danh91 — CHANGES_REQUESTED** [review]
@Ftayri check these suggestions for the issue with the shipment response parsing and the cancel request

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py
Comment (Ftayri) [reply sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py]: I also changed `shipment = lib.to_object(resultMultiParcelValue, response)` to ` shipment = lib.find_element("resultMultiParcelValue", response, resultMultiParcelValue, first=True)` in `_extract_details()` and that fixed it.

**Comment by Ftayri** [comment]
With that, all the current tests are validated! I'll see if I can get a successful shipment cancel response and add it to the tests. I'll also double check the products and services of Chronopost to make sure quickCost is the right request for fetching rates.

File: sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py
Comment (danh91) [reply sdk/extensions/chronopost/karrio/providers/chronopost/shipment/create.py]: 🤔 I made a mistake actually. there was no `result` variable it was supposed to be `shipment_nodes`.

Once that is done, I don't think you need to do this change: 

> I also changed `shipment = lib.to_object(resultMultiParcelValue, response)` to ` shipment = lib.find_element("resultMultiParcelValue", response, resultMultiParcelValue, first=True)` in `_extract_details()` and that fixed it.

File: sdk/extensions/chronopost/tests/chronopost/shipment.py
Comment (danh91) [inline sdk/extensions/chronopost/tests/chronopost/shipment.py:122]: This is still incorrect

File: sdk/extensions/chronopost/tests/chronopost/shipment.py
Comment (danh91) [inline sdk/extensions/chronopost/tests/chronopost/shipment.py:132]: And that one too.

Comment (Ftayri) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py:132]: Can you verify the commits again? I think that you commented on an outdated commit 🤔

File: sdk/extensions/chronopost/tests/chronopost/shipment.py
Comment (danh91) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py]: My bad 👍🏿

File: sdk/extensions/chronopost/tests/chronopost/shipment.py
Comment (danh91) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py]: my bad. Thanks

File: sdk/extensions/chronopost/tests/chronopost/shipment.py
Comment (danh91) [inline sdk/extensions/chronopost/tests/chronopost/shipment.py:122]: actually, the label is still missing

Comment (Ftayri) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py:122]: For some reason the parsed response tag for the label was `pdfetiquette` instead of `pdfEtiquette` 😆 , now there is a problem with serializing the shipment details. I'm getting a circular reference detected that's definitely coming from the label. This is what the parsed response looks like before serializing it. `(ShipmentDetails(carrier_name='chronopost', carrier_id='chronopost', tracking_number='XP696982485FR', shipment_identifier='XP696982485FR', docs=Documents(label=b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00\x00\x01\x08\x02\x00\x00\x00\x90wS\xde\x00\x00\x00\x0cIDATx^c\x98\xc9\xf0\x1f\x00\x02\xce\x01\x99[H\x07\x99\x00\x00\x00\x00IEND\xaeB`\x82', invoice=None), selected_rate=None, label_type=None, meta={}, id=None), [])`

Comment (danh91) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py:122]: I think it is returning a base64 bytes. So you need to encode it to sting

```
label = base64.b64encode(shipment.pdfetiquette).decode("utf-8")
```

Comment (Ftayri) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py:122]: I thought so too, so labels from other carriers are always string? 🤔

Comment (danh91) [reply sdk/extensions/chronopost/tests/chronopost/shipment.py]: They all come as a string but the XML parser will sometime convert it into bytes.

**Comment by Ftayri** [comment]
I update rating request to use `calculateProducts`, `quickCost` actually returns a bunch of shipping options, to supplement the service, that are not used anymore. I also started implementing tracking. I'll have to find a working tracking response to be able to implement it correctly, the test account I'm using can't do that.

**Comment by Ftayri** [comment]
@danh91 I'm having trouble with parsing tracking response, the schema is weird and I was relying on an actual envelope response to implement the parsing but it's not working, I'm only getting the carrier settings in the parsed response and none of the tracking details `[[], [{'carrier_id': 'chronopost', 'carrier_name': 'chronopost'}]]`

**Comment by Ftayri** [comment]
I managed to parse all of the details except the skybillnumber 😅

**Comment by Ftayri** [comment]
Okay I managed to fix the issue, though I'm not sure if my implementation is clean so I'd appreciate your feedback!

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:10]:
```suggestion
```

I would suggest using `lib` instead because it has all these helpers and it is better typed.

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:55]:
```suggestion
        return lib.Deserializable(response, lib.to_element)
```
You can simplify it like this because `lib.to_element` function will bundle the XML responses under one element `<bundle>....</bundle>`.

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:11]: since responses have been bundled together, you will get a single parent element for all

```suggestion
    responses: lib.Element,
```

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:17]: I would suggest getting the `listEventInfoComp` node instead because it is specific to one `skybillnumber` and gives you the events too.
```suggestion
    tracking_nodes: typing.List[chronopost.eventInfoComp] = [
        lib.find_element("listEventInfoComp", response, chronopost.eventInfoComp)
        for response in responses
    ]
```

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:20]: You don't need the wrapping list. since it will the parse error function will retrieve the list of errors
```suggestion
    errors = provider_error.parse_error_response(result, settings)
```

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:24]: Once you fix the selection above to get the list of `listEventInfoComp` instead, you can just pass that here
```suggestion
    tracking_details = [
        _extract_details(node, settings) for node in tracking_nodes
    ]
```

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:33]: With the fixes above, all you will need is the node and settings as parameters
```suggestion
def _extract_details(
    detail: chronopost.eventInfoComp,
    settings: provider_utils.Settings,
) -> models.TrackingDetails:
```

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:36]: This won't be necessary anymore.

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:53]: Now, you should be able to do
```suggestion
    return models.TrackingDetails(
        carrier_id=settings.carrier_id,
        carrier_name=settings.carrier_name,
        tracking_number=detail.skybillNumber,
        events=[
            models.TrackingEvent(
                date=lib.fdate(event.eventDate, "%Y-%m%dT%H%:%M:%S"),
                description=event.eventLabel,
                code=event.code,
                time=lib.ftime(event.eventDate, "%Y-%m%dT%H%:%M:%S"),
                location=event.officeLabel,
            )
            for event in detail.events
        ],
        delivered=any([event.code == "D" for event in detail.events]),
    )
```

File: sdk/extensions/chronopost/tests/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/tests/chronopost/tracking.py:130]: This should just be 
```suggestion
    [],
```

Fixing the error parsing above should correct that

File: sdk/extensions/chronopost/tests/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/tests/chronopost/tracking.py:45]: I would recommend adding a the test for tracking error parsing.

File: sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/mappers/chronopost/proxy.py:53]: You can replace `exec_parrallel` by
```suggestion
        response: typing.List[str] = lib.run_concurently(get_tracking, request.serialize())
```

**Review by danh91 — CHANGES_REQUESTED** [review]
Great work. You are almost there.
I left a couple of comment for the tracking integration

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:36]: This won't be necessary anymore
```suggestion

```

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:62]: Is the Header required?

File: sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py
Comment (Ftayri) [reply sdk/extensions/chronopost/karrio/providers/chronopost/tracking.py:17]: I was going with this approach at first, but when I tried to get an attribute like eventDate from listEventInfoComps.events, nothing shows up in the editor. It lured me into thinking I was using the wrong classes 😅 but apparently it works.

File: sdk/extensions/chronopost/tests/chronopost/tracking.py
Comment (Ftayri) [reply sdk/extensions/chronopost/tests/chronopost/tracking.py:45]: Okay, I wanted to leave parsing errors for last because it's kind of weird. All chronopost responses are in a `<return></return>`  tag and in that tag you always find `errorCode` which is 0 if everything is fine and `errorMessage` that is sent when an error occures. I couldn't figure out how to parse this.

**Comment by Ftayri** [comment]
I changed what's needed in tracking and changed `parse_error_response` in provider error to be able to correctly parse the errors. thought I'm not sure if it's the optimal way to do it 🤔 I also added error responses for all of the tests and validated them. As always I would greatly appreciate your input!

File: sdk/extensions/chronopost/karrio/providers/chronopost/error.py
Comment (danh91) [inline sdk/extensions/chronopost/karrio/providers/chronopost/error.py:30]: This works. Here is a proposition. to make it lighter,
The good thing about the XML parsing library is that if it is fail-safe and will ignore nodes it doesn't know and process the one it recognizes.

So you don't need to know if it is a `shippingMultiParcelV5Response` or `calculateProductsResponse`. 
Either class will be able to parse all the responses and collect the errorCode and errorMessage since those are the fields they have in common.

Therefore, we just need to take one class that knows how to parse a `<return>...</return>` node and used it on all the responses to parse and check if they have an `errorCode` different from 0.

Please try this and let me know.

```suggestion
import typing
from karrio.core.models import Message
from karrio.providers.chronopost.utils import Settings
import chronopost_lib.shippingservice as ups
import karrio.lib as lib

ReturnType = ups. resultMultiParcelExpeditionValue

def parse_error_response(response: lib.Element, settings: Settings) -> typing.List[Message]:
    errors = lib.find_element("return", response, ReturnType)
    
    return [
        Message(
            carrier_id=settings.carrier_id,
            carrier_name=settings.carrier_name,
            message=error.errorMessage,
            code=error.errorCode,
        )
        for error in errors
        if error.errorCode != 0
    ]
```

**Comment by danh91** [comment]
Overall, fantastic job looks pretty much ready to merge.

File: sdk/extensions/chronopost/karrio/providers/chronopost/error.py
Comment (Ftayri) [reply sdk/extensions/chronopost/karrio/providers/chronopost/error.py:30]: Mhm, this is definitely better!

**Comment by Ftayri** [comment]
@danh91 just waiting for your approval 😃

**Review by danh91 — APPROVED** [review]

**Comment by danh91** [comment]
Hi @Ftayri,

Few final things for the merge.
- please run `black` on `sdk/extensions/chronopost` (just the extension not the schema)
- rebase your branch on the `upstream karrio-2022.8` release branch
- Then let me know and I will merge

Very excited. Thank you

**Comment by Ftayri** [comment]
Just to confirm, I should `git fetch upstream` then `git checkout karrio-2022.8` and `git rebase upstream/karrio-2022.8` right? 😅 I'm worrying too much.

**Comment by danh91** [comment]
> Just to confirm, I should `git fetch upstream` then `git checkout karrio-2022.8` and `git rebase upstream/karrio-2022.8` right? 😅 I'm worrying too much.

Actually, just run black. I will take care of the rebase.

**Comment by Ftayri** [comment]
Already done.

**Pull request merged**
